crates/red_knot_module_resolver/src/path.rs

```diff
@@ -0,0 +1,693 @@
+                path.with_extension("py"),
+            ))),
+            Self::FirstParty(FirstPartyPath(path)) => Some(ModuleResolutionPath::FirstParty(
+                FirstPartyPathBuf(path.with_extension("py")),
+            )),
+            Self::StandardLibrary(_) => None,
+            Self::SitePackages(SitePackagesPath(path)) => Some(ModuleResolutionPath::SitePackages(
+                SitePackagesPathBuf(path.with_extension("py")),
+            )),
+        }
+    }
+
+    #[cfg(test)]
+    #[must_use]
+    pub(crate) fn to_module_resolution_path(self) -> ModuleResolutionPath {
+        match self {
+            Self::Extra(ExtraPath(path)) => {
+                ModuleResolutionPath::Extra(ExtraPathBuf(path.to_path_buf()))
+            }
+            Self::FirstParty(FirstPartyPath(path)) => {
+                ModuleResolutionPath::FirstParty(FirstPartyPathBuf(path.to_path_buf()))
+            }
+            Self::StandardLibrary(StandardLibraryPath(path)) => {
+                ModuleResolutionPath::StandardLibrary(StandardLibraryPathBuf(path.to_path_buf()))
+            }
+            Self::SitePackages(SitePackagesPath(path)) => {
+                ModuleResolutionPath::SitePackages(SitePackagesPathBuf(path.to_path_buf()))
+            }
+        }
+    }
+
+    #[cfg(test)]
+    #[must_use]
+    pub(crate) fn join(
+        &self,
+        component: &'a (impl AsRef<FileSystemPath> + ?Sized),
+    ) -> ModuleResolutionPath {
+        let mut result = self.to_module_resolution_path();
+        result.push(component.as_ref().as_str());
+        result
+    }
+
+    #[must_use]
+    #[inline]
+    fn as_file_system_path(self) -> &'a FileSystemPath {
+        match self {
+            Self::Extra(ExtraPath(path)) => path,
+            Self::FirstParty(FirstPartyPath(path)) => path,
+            Self::StandardLibrary(StandardLibraryPath(path)) => path,
+            Self::SitePackages(SitePackagesPath(path)) => path,
+        }
+    }
+}
+
+impl<'a> From<&'a ModuleResolutionPath> for ModuleResolutionPathRef<'a> {
+    #[inline]
+    fn from(value: &'a ModuleResolutionPath) -> Self {
+        match value {
+            ModuleResolutionPath::Extra(ExtraPathBuf(path)) => {
+                ModuleResolutionPathRef::extra_unchecked(path)
+            }
+            ModuleResolutionPath::FirstParty(FirstPartyPathBuf(path)) => {
+                ModuleResolutionPathRef::first_party_unchecked(path)
+            }
+            ModuleResolutionPath::StandardLibrary(StandardLibraryPathBuf(path)) => {
+                ModuleResolutionPathRef::standard_library_unchecked(path)
+            }
+            ModuleResolutionPath::SitePackages(SitePackagesPathBuf(path)) => {
+                ModuleResolutionPathRef::site_packages_unchecked(path)
+            }
+        }
+    }
+}
+
+impl<'a> AsRef<FileSystemPath> for ModuleResolutionPathRef<'a> {
+    #[inline]
+    fn as_ref(&self) -> &FileSystemPath {
+        self.as_file_system_path()
+    }
+}
```

Comment:
Do we rely on this logic a lot. I'm worried that this will bite us when adding support for vendored file system because implementing `AsRef` then becomes impossible.

Reply:
I've just made a number of details more private to `path.rs`, so that we don't rely on this logic nearly as much. I think it should be fairly easy to adjust it when we add support for vendored files.

Reply:
While writing tests, I realised that the two places where we still used this... were in fact bugs. So it's gone entirely now 😄